File: osd32mp1/rt_jitter.c
/* Measure scheduling lateness at flight-loop rates. BUILDS AND RUNS ON THE BOARD.
 *
 * The question: FreeRTOS's Posix port makes each task a pthread, so every task
 * wake-up inherits the host scheduler's latency. OpenPilot's
 * stabilizationInnerloopTask runs at PIOS_SENSOR_RATE (500 Hz = 2 ms) and its
 * own watchdog treats 2 missed gyro updates as a warning and 3 as critical.
 * So "3 periods late" is the firmware's own definition of a critical event -
 * that is the threshold reported here.
 *
 * This is C, not Python, on purpose: the flight code is C, and a Python
 * interpreter's own overhead would be indistinguishable from scheduler latency.
 *
 * clock_nanosleep(TIMER_ABSTIME) against CLOCK_MONOTONIC is what a correctly
 * written realtime loop does - it does not accumulate drift, so everything left
 * over IS the host's scheduling error.
 *
 *   cc -O2 -o rt_jitter rt_jitter.c -lpthread
 *   ./rt_jitter 500 20 0            # 500 Hz, 20 s, no extra load
 *   chrt -f 50 ./rt_jitter 500 20 0 # same under SCHED_FIFO
 *   ./rt_jitter 500 20 2            # with 2 busy cores of contention
 */
#define _GNU_SOURCE
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include <unistd.h>
#include <sched.h>
#include <sys/mman.h>
#include <sys/wait.h>

static int cmp_d(const void *a, const void *b)
{
    double x = *(const double *)a, y = *(const double *)b;
    return (x > y) - (x < y);
}

static double pct(double *sorted, int n, double p)
{
    int k = (int)(p / 100.0 * (n - 1) + 0.5);
    if (k < 0) k = 0;
    if (k >= n) k = n - 1;
    return sorted[k];
}

int main(int argc, char **argv)
{
    double hz      = argc > 1 ? atof(argv[1]) : 500.0;
    double seconds = argc > 2 ? atof(argv[2]) : 20.0;
    int    nload   = argc > 3 ? atoi(argv[3]) : 0;

    /* A page fault in the flight loop is indistinguishable from a scheduling
     * delay, and just as fatal. Real RT code locks its memory down; do the
     * same here so we measure the scheduler and not the pager. */
    int locked = (mlockall(MCL_CURRENT | MCL_FUTURE) == 0);

    pid_t kids[16];
    int nkids = 0;
    for (int i = 0; i < nload && i < 16; i++) {
        pid_t p = fork();
        if (p == 0) {
            /* fork() inherits the scheduling policy. Left alone, a busy child
             * would run at the parent's SCHED_FIFO priority and - FIFO being
             * run-to-completion - starve the very loop we are measuring until
             * the kernel's RT throttle intervened. The realistic case is flight
             * code at RT priority and everything else normal, so force that. */
            struct sched_param z = { .sched_priority = 0 };
            sched_setscheduler(0, SCHED_OTHER, &z);
            struct timespec e;
            clock_gettime(CLOCK_MONOTONIC, &e);
            double end = e.tv_sec + e.tv_nsec / 1e9 + seconds + 2;
            volatile unsigned x = 0;
            for (;;) {
                for (int j = 0; j < 200000; j++) x = x * 1103515245u + 12345u;
                clock_gettime(CLOCK_MONOTONIC, &e);
                if (e.tv_sec + e.tv_nsec / 1e9 > end) _exit(0);
            }
        }
        kids[nkids++] = p;
    }

    long period_ns = (long)(1e9 / hz);
    int  n = (int)(seconds * hz);
    double *late = malloc(sizeof(double) * n);
    if (!late) return 1;

    struct timespec t;
    clock_gettime(CLOCK_MONOTONIC, &t);
    for (int i = 0; i < n; i++) {
        t.tv_nsec += period_ns;
        while (t.tv_nsec >= 1000000000L) { t.tv_nsec -= 1000000000L; t.tv_sec++; }
        clock_nanosleep(CLOCK_MONOTONIC, TIMER_ABSTIME, &t, NULL);
        struct timespec now;
        clock_gettime(CLOCK_MONOTONIC, &now);
        late[i] = ((now.tv_sec - t.tv_sec) * 1e9 + (now.tv_nsec - t.tv_nsec)) / 1e6;
    }

    for (int i = 0; i < nkids; i++) waitpid(kids[i], NULL, 0);

    double period_ms = period_ns / 1e6;
    double sum = 0, worst = 0;
    int over1 = 0, over3 = 0;
    for (int i = 0; i < n; i++) {
        sum += late[i];
        if (late[i] > worst) worst = late[i];
        if (late[i] > period_ms) over1++;
        if (late[i] > 3 * period_ms) over3++;
    }
    qsort(late, n, sizeof(double), cmp_d);

    int pol = sched_getscheduler(0);
    const char *pn = pol == SCHED_FIFO ? "SCHED_FIFO" :
                     pol == SCHED_RR   ? "SCHED_RR"   : "SCHED_OTHER";
    struct sched_param sp;
    sched_getparam(0, &sp);

    printf("=== %.0f Hz (%.2f ms period), %d wake-ups, %s prio %d, load=%d, mlock=%s ===\n",
           hz, period_ms, n, pn, sp.sched_priority, nload, locked ? "yes" : "NO");
    printf("  mean late   %8.3f ms\n", sum / n);
    printf("  median      %8.3f ms\n", pct(late, n, 50));
    printf("  p99         %8.3f ms\n", pct(late, n, 99));
    printf("  p99.9       %8.3f ms\n", pct(late, n, 99.9));
    printf("  WORST       %8.3f ms   = %.1f periods\n", worst, worst / period_ms);
    printf("  late by >1 period : %6d  (%.3f %%)\n", over1, 100.0 * over1 / n);
    printf("  late by >3 periods: %6d  (%.3f %%)   <- firmware calls this critical\n",
           over3, 100.0 * over3 / n);
    free(late);
    return 0;
}

File: osd32mp1/rt_jitter.py
#!/usr/bin/env python3
"""Measure scheduling jitter at flight-loop rates. RUNS ON THE BOARD.

The question this answers: if FreeRTOS is a set of pthreads on Linux, how late
does a task actually wake up? OpenPilot's stabilizationInnerloopTask is
gyro-triggered at PIOS_SENSOR_RATE (500 Hz = 2 ms), and its own watchdog wants
exactly ONE gyro update per run - it warns at 2 and calls 3 critical. So a
wake-up 3 periods late is, by the firmware's own definition, a critical event.

What is measured is the LATENESS of a periodic wake-up: ask to sleep until
t0+n*period, then see what the clock really says. That is exactly the error a
FreeRTOS tick or a task wake-up inherits from the host scheduler.

Run it at normal priority and again under SCHED_FIFO to see what RT priority
buys:

    ./rt_jitter.py --hz 500 --seconds 20
    chrt -f 50 ./rt_jitter.py --hz 500 --seconds 20
    ./rt_jitter.py --hz 500 --seconds 20 --load 2     # with CPU contention
"""
import argparse
import os
import sys
import time


def percentile(xs, p):
    if not xs:
        return 0.0
    s = sorted(xs)
    k = min(len(s) - 1, max(0, int(round((p / 100.0) * (len(s) - 1)))))
    return s[k]


def main():
    ap = argparse.ArgumentParser()
    ap.add_argument("--hz", type=float, default=500.0)
    ap.add_argument("--seconds", type=float, default=20.0)
    ap.add_argument("--load", type=int, default=0,
                    help="spawn N busy processes to create contention")
    a = ap.parse_args()

    kids = []
    for _ in range(a.load):
        pid = os.fork()
        if pid == 0:
            end = time.time() + a.seconds + 2
            x = 0
            while time.time() < end:
                x = (x * 1103515245 + 12345) & 0xFFFFFFFF
            os._exit(0)
        kids.append(pid)

    period = 1.0 / a.hz
    n = int(a.seconds * a.hz)
    late = []
    t0 = time.perf_counter()
    for i in range(1, n + 1):
        target = t0 + i * period
        now = time.perf_counter()
        if target > now:
            time.sleep(target - now)
        late.append((time.perf_counter() - target) * 1000.0)   # ms late

    for pid in kids:
        try:
            os.waitpid(pid, 0)
        except OSError:
            pass

    try:
        pol = os.sched_getscheduler(0)
        polname = {0: "SCHED_OTHER", 1: "SCHED_FIFO", 2: "SCHED_RR"}.get(pol, str(pol))
        prio = os.sched_getparam(0).sched_priority
    except Exception:
        polname, prio = "?", "?"

    mean = sum(late) / len(late)
    print(f"=== {a.hz:.0f} Hz ({period*1000:.2f} ms period), {len(late)} wake-ups, "
          f"{polname} prio {prio}, load={a.load} ===")
    print(f"  mean late   {mean:8.3f} ms")
    print(f"  median      {percentile(late,50):8.3f} ms")
    print(f"  p99         {percentile(late,99):8.3f} ms")
    print(f"  p99.9       {percentile(late,99.9):8.3f} ms")
    print(f"  WORST       {max(late):8.3f} ms   = {max(late)/(period*1000):.1f} periods")
    over = [x for x in late if x > period * 1000]
    print(f"  wake-ups later than one full period: {len(over)} "
          f"({100.0*len(over)/len(late):.2f} %)")
    # The firmware's own thresholds: >1 gyro update missed = warn, >3 = critical
    crit = [x for x in late if x > 3 * period * 1000]
    print(f"  later than THREE periods (firmware calls this critical): {len(crit)} "
          f"({100.0*len(crit)/len(late):.3f} %)")
    return 0


if __name__ == "__main__":
    sys.exit(main())
